src/terrain/generation/biome.cpp

```diff
@@ -98,6 +135,108 @@ Biome::Biome(const std::string& biome_name) {
     }
 }
 
+void
+Biome::read_tile_macro_data(const Json::Value& biome_data) {
+    // for tile macro in data biome
+    for (const Json::Value& tile_macro : biome_data["Tile_Macros"]) {
+        // create a land generator for each tile macro
+        generation::LandGenerator gen(tile_macro["Land_Data"]);
+        land_generators_.push_back(gen);
+    }
+}
+
+void
+Biome::read_map_tile_data(const Json::Value& biome_data) {
+    // add tile macro to tiles
+    for (const Json::Value& tile_type : biome_data["Tile_Data"]) {
+        std::vector<TileMacro_t> tile_macros;
+        for (const Json::Value& tile_macro_id : tile_type["Land_From"]) {
+            TileMacro_t tile_macro = tile_macro_id.asInt();
+            if (tile_macro >= land_generators_.size()) [[unlikely]] {
```

Comment:
you should used named land generators and an `unordered_map`

Reply:
All land generaotors have an "id" that coresponds to the index in the vector.

Reply:
that's easy to break

Reply:
how? The if statement is right there.

Reply:
whatever.